pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -611,6 +642,24 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, ru
 		}(container)
 	}
 	wg.Wait()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+		wg.Add(len(sidecars))
+		for _, container := range sidecars {
+			go func(container *kubecontainer.Container) {
+				defer utilruntime.HandleCrash()
+				defer wg.Done()
+
+				killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, container.Name)
+				if err := m.killContainer(pod, container.ID, container.Name, "", gracePeriodOverride); err != nil {
```

Comment:
In this implementation, I see that the same termination grace period is used twice; is that intentional? This would result in both shutdown stages having the same grace period. If I understand correctly, with this implementation, a Pod that has the termination grace period set to X actually can terminate for as long as 2X. 

Or should the implementation be such that both stages share part of the termination grace period? i.e. `G` = grace period, `X` = time it took to shutdown the non-sidecar containers, then the sidecar containers are terminated for `G - X`.

Reply:
So from my understanding termination grace period is only used on a per container basis, not on a pod basis, i.e. it's the time each container has to respond to SIGTERM before being sent a SIGKILL. 
In practice this results in the grace period of the containers being the same as the pod due to them all being killed in parallel. By breaking this into two it does mean that time for a pod to terminate could be 2X. Currently nothing (as far as I'm aware) keeps track of the remaining grace period at a pod level so some more work would need to be done to work out how to do this.
I also thought maybe to err on the side of allowing time for everything to have a chance to gracefully shutdown, instead of sending a SIGKILL to all the sidecars if a container isn't behaving.

Reply:
OK, I thought about this a bit and I think the right solution is to share the grace period between the non-sidecars and sidecars. We currently run a version of this patch where we do this. It makes intuitive sense: it would be strange for a user to read in the documentation "terminationGracePeriodSeconds can potentially be 2X if the pod has sidecar containers." It should be up to the user to explicitly specify how long they think their pods need to terminate (combining sidecar time and non-sidecar time). What do you think?

Reply:
right, so I had a play and I have implemented something similar to what you've done, it seems to work as expected in my testing. I've pushed a new commit with this.

Reply:
Thanks! this is looking good.

Reply:
@tomwans before we introduce container kill order, there was the assumption that all containers would get killed roughly at the same time so in production pod graceful termination timeout was set based on longest container termination timeout.

here since we are sharing same grace period, my concern is that the original assumption got broken and can cause some regression

Reply:
@zhan849 yep, that's true, that can be confusing as well. But sharing the grace period between the two stages is more flexible because it allows you to set a tighter grace period. For example, if you have the two stages, S₁ and S₂, if you use the same grace period, you have to make grace period = MAX(S₁, S₂), which results in the real termination time being 2 * MAX(S₁, S₂). Assuming S₁ takes 0 seconds to terminate and S₂ takes 10, it will result in a much longer total grace period than necessary (20). By sharing the termination grace period, we set a tighter window since it'll be S₁ + S₂ = 10.

I think we have two choices:

1) Pick some behavior (Personally, I think sharing the grace period makes the most intuitive sense) and update the docs to be very clear about this behavior change as we roll this out into alpha
2) Introduce a Container-level termination grace period, so each container has its own grace period.

WDYT?

Reply:
@tomwans I agree that it intuitively makes more sense to share the graceful period. However given the fact that sidecars are usually developed independently to applications, I am afraid that it adds up people's concern that "I'm gonna get non-deterministic graceful period because I'm gonna share that period with some container that i don't have control over."

Conservatively speaking I'm more inclined to not have shutdown sequence in alpha (at least not worth than current scenario) and get the feature out first and re-think about the feasibility of adding container-level grace period, without which, we could possibly introduce confusion anyway

Reply:
As a user of a network proxy sidecar pattern, for us shutdown ordering has been trickier than startup ordering. We use sidecars that are developed by a third party and we develop the application containers. For the case of startup, we can just modify our application's entrypoint to sleep or loop waiting on the proxy. On the case of shutdown, we need to both modify the sidecar vendor's image to intercept the interrupt and our application to shutdown the sidecar as a final step to its own shutdown. Thank you all for your work on this feature, I am very thankful to see this progress.